### DISCUSSION TITLE
Texture distortion

### DISCUSSION THREAD

**Discussion by MyShallowTime** [KhronosGroup/WebGL#3565] [Q&A]
Hello, have you encountered this problem?

When I draw this image in small size, part of it disappears.
![point](https://github.com/KhronosGroup/WebGL/assets/20198963/25913a68-d0ca-4d38-8a4d-5db59294c8e8)
![image](https://github.com/KhronosGroup/WebGL/assets/20198963/ccc5d6eb-0281-4aab-b44b-f40d424a6d4e)
Look, the vertical lines disappeared。why？

Does anyone know why?
Thanks in advance!

demo:
[texture.zip](https://github.com/KhronosGroup/WebGL/files/11979231/texture.zip)

**Comment by kdashg** [comment]
Probably just naive texture minification sampling. If you shrink something enough, the two texels averaged for sampling can stretch across the line.
```
        7px: | W | W | W | B | W | W | W |
 Sampled at:   ^    ^   ^     ^   ^    ^
Averages to: | W +  W | W  +  W | W +  W |
```
If you want that visual data to be preserved, you'll want to calculate and supply your own mipmaps.

**Reply by MyShallowTime** [reply]
Thank you very much, I will try it!

**Reply by MyShallowTime** [reply]
Hi, @kdashg 
How to calculate and supply own mipmaps.
I only see this api gl.generateMipmaps(type)， type has only a few fixed values

**Reply by greggman** [reply]
You can upload mips levels by setting the 2nd parameter for `texImage2D` or `texSubImage2` to the mip level you're supplying data for. The 2nd the last example on [this page](https://webgl2fundamentals.org/webgl/lessons/webgl-3d-textures.html) has every mip level in different colors.

As for calculating the mip level data on your own, that's a huge topic. You'd use [an image scaling algorithm](https://en.wikipedia.org/wiki/Image_scaling). WebGL does not provide these. You have to provide them yourself.

**Reply by MyShallowTime** [reply]
Thanks, it feels hard for me. I still need to learn.
